File: pillow-rs/tests/tobytes_oracle.rs
//! Exact Rust API execution of the shared Pillow Image.tobytes corpus.
#![allow(unused_crate_dependencies)]

use std::fs;
use std::path::{Path, PathBuf};

use pillow_rs::image::Image;
use serde::Deserialize;
use serde_json::Value;

#[derive(Deserialize)]
struct InputFixture {
    cases: Vec<InputCase>,
}

#[derive(Deserialize)]
struct InputCase {
    id: String,
    mode: String,
    input: InputSpec,
    params: EncoderParams,
}

#[derive(Deserialize)]
struct InputSpec {
    source: String,
    size: [u32; 2],
    color: Option<Value>,
}

#[derive(Deserialize)]
struct EncoderParams {
    #[serde(rename = "_args", default)]
    args: Vec<String>,
}

#[derive(Deserialize)]
struct OutputFixture {
    cases: Vec<OutputCase>,
}

#[derive(Deserialize)]
struct OutputCase {
    id: String,
    assert: OracleAssertion,
}

#[derive(Deserialize)]
struct OracleAssertion {
    method: String,
    reference: String,
    raw_kind: String,
}

fn repo_root() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .expect("pillow-rs crate has a workspace parent")
        .to_owned()
}

fn constant_color(value: &Value) -> (u8, u8, u8, u8) {
    let values = value.as_array().expect("constant color is an array");
    let component = |index: usize, default: u8| {
        values
            .get(index)
            .and_then(Value::as_u64)
            .map_or(default, |item| item as u8)
    };
    (
        component(0, 0),
        component(1, 0),
        component(2, 0),
        component(3, 255),
    )
}

fn create_input(root: &Path, case: &InputCase) -> Image {
    match case.input.source.as_str() {
        "constant" => Image::new(
            case.input.size[0],
            case.input.size[1],
            &case.mode,
            constant_color(case.input.color.as_ref().expect("constant color exists")),
        )
        .expect("constant image constructs"),
        "reference_rgb" => {
            let source = Image::open_bytes(
                fs::read(root.join("tests/test_reference.png")).expect("reference image exists"),
            )
            .expect("reference image opens");
            source
                .resize((case.input.size[0], case.input.size[1]), Some("LANCZOS"))
                .expect("reference image resizes")
                .convert(&case.mode, None, None, None, None)
                .expect("reference image converts")
        }
        source => panic!("unsupported fixture input source {source}"),
    }
}

#[test]
fn shared_tobytes_corpus_matches_pillow_exactly() {
    let root = repo_root();
    let input: InputFixture = serde_json::from_slice(
        &fs::read(root.join("tests/fixtures/input/jsons/Image.tobytes.json"))
            .expect("Image.tobytes inputs exist"),
    )
    .expect("Image.tobytes inputs parse");
    let output: OutputFixture = serde_json::from_slice(
        &fs::read(root.join("tests/fixtures/outputs/jsons/Image.tobytes.json"))
            .expect("Image.tobytes oracles exist"),
    )
    .expect("Image.tobytes oracles parse");

    assert_eq!(input.cases.len(), 8);
    for case in input.cases {
        let assertion = &output
            .cases
            .iter()
            .find(|candidate| candidate.id == case.id)
            .unwrap_or_else(|| panic!("{}: oracle exists", case.id))
            .assert;
        assert_eq!(assertion.method, "image", "{}: assertion method", case.id);
        assert_eq!(
            assertion.raw_kind, "bytes",
            "{}: exact byte assertion",
            case.id
        );

        let image = create_input(&root, &case);
        let actual = if case.params.args.is_empty() {
            image.tobytes().expect("public bytes materialize")
        } else {
            image
                .tobytes_encoded(&case.mode, &case.params.args[0], &case.params.args[1..])
                .expect("encoded public bytes materialize")
        };
        assert_eq!(
            actual,
            fs::read(
                root.join("tests/fixtures/outputs")
                    .join(&assertion.reference)
            )
            .expect("raw Pillow oracle exists"),
            "{}: exact public bytes",
            case.id
        );
    }
}

File: deprecated/migration-parity-v0/wasm/tests/tobytes_oracle.mjs
import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

import { loadOracleCorpus } from './oracle_corpus.mjs';

const here = dirname(fileURLToPath(import.meta.url));
const repoRoot = join(here, '..', '..');
const packageDir = join(here, '..', 'pkg', 'core');
const wasm = readFileSync(join(packageDir, 'pillow_rs_js_bg.wasm'));
const bindings = await import(pathToFileURL(join(packageDir, 'pillow_rs_js.js')));
bindings.initSync({ module: new WebAssembly.Module(wasm) });

function createInput(testCase) {
    const spec = testCase.input.input;
    if (spec.source === 'constant') {
        const values = Array.isArray(spec.color)
            ? [...spec.color]
            : [spec.color, 0, 0, 0];
        while (values.length < 4) values.push(values.length === 3 ? 255 : 0);
        return new bindings.Image(
            testCase.input.mode,
            spec.size[0],
            spec.size[1],
            values[0],
            values[1],
            values[2],
            values[3],
        );
    }
    if (spec.source !== 'reference_rgb') {
        throw new Error(`${testCase.id}: unsupported input source ${spec.source}`);
    }
    const source = bindings.Image.open(
        readFileSync(join(repoRoot, 'tests', 'test_reference.png')),
    );
    return source
        .resize(spec.size[0], spec.size[1], 'LANCZOS')
        .convert(testCase.input.mode);
}

const cases = loadOracleCorpus().cases.filter(
    (item) => item.suite === 'fixtures'
        && item.operation.module === 'Image'
        && item.operation.target === 'tobytes',
);

if (cases.length !== 8) {
    throw new Error(`Image.tobytes corpus selected ${cases.length} cases; expected 8`);
}

for (const testCase of cases) {
    const assertion = testCase.assertion;
    if (assertion.method !== 'image' || assertion.raw_kind !== 'bytes') {
        throw new Error(`${testCase.id}: expected exact public byte assertion`);
    }
    const image = createInput(testCase);
    const positional = testCase.input.params._args ?? [];
    const actual = positional.length === 0
        ? image.toBytes()
        : image.toBytesEncoded(positional[0], positional.slice(1));
    const expected = readFileSync(join(testCase.outputRoot, assertion.reference));
    if (!Buffer.from(actual).equals(expected)) {
        const bytes = Buffer.from(actual);
        const first = bytes.findIndex((value, index) => value !== expected[index]);
        throw new Error(
            `${testCase.id}: exact bytes differ at ${first}`
            + ` (${bytes[first]} != ${expected[first]})`,
        );
    }
}

process.stdout.write(
    `core: ${cases.length} exact Image.tobytes Pillow oracle cases passed\n`,
);
